tsfc/driver.py

```diff
@@ -346,7 +365,59 @@ def compile_expression_dual_evaluation(expression, to_element, coordinates, *,
                       index_cache={},
                       scalar_type=parameters["scalar_type"])
 
-    if all(isinstance(dual, PointEvaluation) for dual in to_element.dual_basis()):
+    elements_with_basis = (FiatElement, TensorFiniteElement, TensorProductElement,
+                           EnrichedElement, FlattenedDimensions, DiscontinuousElement)
+    if isinstance(element, elements_with_basis):
```

Comment:
Could this be generalised without causing issues to
```
    try:
        if isinstance(element.dual_basis, tuple):
          has_dual_basis = True
    except (AttributeError, NotImplementedError):
        has_dual_basis = False
    if has_dual_basis:
```
or is there a good reason for this to be maintained as a list?

Reply:
These isinstance checks are an anti-pattern. I agree with Miklos' refactoring suggestion above.